utils/compression/gzip_compressor.go

```diff
@@ -41,37 +32,32 @@ func (g *gzipCompressor) Compress(msg []byte) ([]byte, error) {
 		return nil, fmt.Errorf("%w: (%d) > (%d)", ErrMsgTooLarge, len(msg), g.maxSize)
 	}
 
-	g.lock.Lock()
-	defer g.lock.Unlock()
+	var writeBuffer bytes.Buffer
+	gzipWriter := g.gzipWriterPool.Get().(*gzip.Writer)
+	gzipWriter.Reset(&writeBuffer)
+	defer g.gzipWriterPool.Put(gzipWriter)
 
-	g.writeBuffer.Reset()
-	g.gzipWriter.Reset(g.writeBuffer)
-	if _, err := g.gzipWriter.Write(msg); err != nil {
+	if _, err := gzipWriter.Write(msg); err != nil {
 		return nil, err
 	}
-	if err := g.gzipWriter.Close(); err != nil {
+	if err := gzipWriter.Close(); err != nil {
 		return nil, err
 	}
-
-	compressed := g.writeBuffer.Bytes()
-	compressedCopy := slices.Clone(compressed)
-	return compressedCopy, nil
+	return writeBuffer.Bytes(), nil
 }
 
 // Decompress decompresses [msg].
 func (g *gzipCompressor) Decompress(msg []byte) ([]byte, error) {
-	g.lock.Lock()
-	defer g.lock.Unlock()
-
-	g.bytesReader.Reset(msg)
-	if err := g.gzipReader.Reset(g.bytesReader); err != nil {
+	bytesReader := bytes.NewReader(msg)
+	gzipReader, err := gzip.NewReader(bytesReader)
```

Comment:
seems we could pool the `gzipReader` too.

Reply:
We benchmarked it - the overhead of the sync pool for the reader is more than the savings for the allocation.